tsnet/tsnet.go

```diff
@@ -518,6 +518,7 @@ func (s *Server) APIClient() (*tailscale.Client, error) {
 
 // Listen announces only on the Tailscale network.
 // It will start the server if it has not been started yet.
+// The addr should not include a hostname, only a port, for example ":80".
 func (s *Server) Listen(network, addr string) (net.Listener, error) {
 	switch network {
 	case "", "tcp", "tcp4", "tcp6":
```

Comment:
will 3 of these 4 be a problem too?

Reply:
That's a good question. The [docs for net.Listen](https://pkg.go.dev/net#Listen) say that you can use a hostname with "tcp4" and "tcp6", so I do not think these values make the situation any worse. The real problem is that if we set the hostname portion of the address to anything other than "", the routing won't work.

But then, if we want to check for "valid" values at the point where we're routing, we'd have to re-parse out the address. Mechanically that's fine, but do we want to spend the resources to do that on each connection? I'm happy to do the work if you think it's worthwhile.

Reply:
Am I correct in thinking that this listener won't even receive a connection if it didn't already arrive with an address wireguard thinks corresponds to this TS instance?

Reply:
I guess as it's written now, we do need them to match because we include it in the key.

It doesn't look like we necessarily need the network type in the key, though—for forwarding purposes, I think we've already asked netstack to only do this for TCP (cf. "ForwardTCPIn, if non-nil, handles forwarding an inbound TCP connection"). 

Could we do the check in Listen to give a sensible response (as now) but then omit the network from the listenKey entirely?